grafana/explore-profiles
# feat(FlameGraph): Keep focus whenever the profile data changes

### ✨ Description

**Related issue(s):** resolves https://github.com/grafana/explore-profiles/issues/260, related PR in Grafana https://github.com/grafana/grafana/pull/98356

This PR introduces a new `keepFocusOnDataChange` prop to the `FlameGraph` component.

### 📖 Summary of the changes

In order to simplify the implementation of the fix, the whole [@grafana/flamegraph](https://github.com/grafana/grafana/tree/main/packages/grafana-flamegraph) package has been copy-pasted. Once agreed on the definitive solution, I'll update [the PR in the Grafana repository](https://github.com/grafana/grafana/pull/98356).

I've commented the changes the diff tab (mainly in `src/tmp/grafana-flamegraph/src/FlameGraphContainer.tsx`).

### 🧪 How to test?

- Manually:
  - Go to the "Flame graph" or the "Diff flame graph" view
  - Click on a flame graph node
  - Click on "Focus block"
  - Change the time range or click on the time picker refresh button
  - The block should stay focused and the percentage should update or drop to 0 if the function didn't consume any resource for the new time range

The same should occur for the sandwich view
